### PR TITLE
RAC-391: add is blacklisted query

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Review by StevenVAIDIE — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Resources/translations/validators.en_US.yml [hunk 1]
```diff
@@ -70,6 +70,7 @@ pim_catalog:
         family_variant_maximum_number_of_level: Family variant cannot have more than "%level%" level
         family_variant_unique_attributes_in_last_level: Unique attribute "%attribute%" must be set at the product level
         product_identifier: These identifiers "%s" don't exist.
+        blacklisted_attribute_code: The attribute code "attribute_code" was previously existing and is currently being cleaned up
```
[the remark this answers is not in the document]
Comment (juliensnz) [reply src/Akeneo/Pim/Enrichment/Bundle/Resources/translations/validators.en_US.yml:73]: Nice, you are right

File: src/Akeneo/Pim/Structure/Bundle/Resources/config/validators.yml [hunk 2]
```diff
@@ -87,6 +87,13 @@ services:
         tags:
             - { name: validator.constraint_validator, alias: pim_single_identifier_attribute_validator }
 
+    pim_catalog.validator.constraint.blaclisted_attribute_code:
```
Comment (ValentinMumble) [inline src/Akeneo/Pim/Structure/Bundle/Resources/config/validators.yml:90]:
```suggestion
    pim_catalog.validator.constraint.blacklisted_attribute_code:
```

**Review by ValentinMumble — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Resources/translations/validators.en_US.yml [hunk 3]
```diff
@@ -70,6 +70,7 @@ pim_catalog:
         family_variant_maximum_number_of_level: Family variant cannot have more than "%level%" level
         family_variant_unique_attributes_in_last_level: Unique attribute "%attribute%" must be set at the product level
         product_identifier: These identifiers "%s" don't exist.
+        blacklisted_attribute_code: The attribute code "%attribute_code%" was previously existing and is currently being cleaned up
```
Comment (SamirBoulil) [inline src/Akeneo/Pim/Enrichment/Bundle/Resources/translations/validators.en_US.yml:73]: I think this message needs english validation, to see with @felix-couprie or @camillebourjade

File: tests/back/Pim/Structure/Integration/Attribute/Query/IsAttributeCodeBlacklistedIntegration.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+namespace AkeneoTest\Pim\Structure\Integration\Family;
+
+use Akeneo\Test\Integration\Configuration;
+use Akeneo\Test\Integration\TestCase;
+
+final class IsAttributeCodeBlacklistedIntegration extends TestCase
+{
+    public function test_it_check_if_existing_attribute_code_is_blacklisted(): void
+    {
+        $query = $this->get('akeneo.pim.structure.query.is_attribute_code_blacklisted');
```
Comment (SamirBoulil) [inline tests/back/Pim/Structure/Integration/Attribute/Query/IsAttributeCodeBlacklistedIntegration.php:14]: not mandatory at all, you could instanciate the $query in the setup function

File: tests/back/Pim/Structure/Integration/Attribute/Query/IsAttributeCodeBlacklistedIntegration.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        $result = $query->execute('description');
+
+        $this->assertFalse($result);
+    }
+
+    public function test_it_check_if_blacklisted_attribute_code_is_blacklisted(): void
+    {
+        $query = $this->get('akeneo.pim.structure.query.is_attribute_code_blacklisted');
+
+        $blacklistAttributeCode = <<<SQL
```
Comment (SamirBoulil) [inline tests/back/Pim/Structure/Integration/Attribute/Query/IsAttributeCodeBlacklistedIntegration.php:24]: I'd extract this bit in a `private blacklistAttribute($attributeCode)`

File: tests/back/Pim/Structure/Integration/Attribute/Validation/BlacklistedAttributeCodeIntegration.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace AkeneoTest\Pim\Structure\Integration\Attribute\Validation;
+
+class BlackListedCodeIntegration extends AbstractAttributeTestCase
+{
+    public function testAttributeIsNotBlacklisted()
```
Comment (SamirBoulil) [inline tests/back/Pim/Structure/Integration/Attribute/Validation/BlacklistedAttributeCodeIntegration.php:7]:
```suggestion
    public function test_I_can_update_an_attribute_that_is_not_blacklisted()
```

File: tests/back/Pim/Structure/Integration/Attribute/Validation/BlacklistedAttributeCodeIntegration.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                'type'             => 'pim_catalog_text',
+                'group'            => 'attributeGroupA'
+            ]
+        );
+
+        $violations = $this->validateAttribute($attribute);
+
+        $this->assertCount(0, $violations);
+    }
+    public function testAttributeIsBlacklisted()
```
Comment (SamirBoulil) [inline tests/back/Pim/Structure/Integration/Attribute/Validation/BlacklistedAttributeCodeIntegration.php:24]:
```suggestion
    public function test_I_cannot_update_an_blacklisted_attribute()
```

File: tests/back/Pim/Structure/Integration/Attribute/Validation/BlacklistedAttributeCodeIntegration.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        $blacklistAttributeCode = <<<SQL
+        INSERT INTO `pim_catalog_attribute_blacklist` (`attribute_code`)
+        VALUES
+            ('new_blacklisted_attribute');
+        SQL;
+
+        /** @var Connection $connection */
+        $connection = $this->get('database_connection');
+        $connection->executeUpdate($blacklistAttributeCode);
```
Comment (SamirBoulil) [inline tests/back/Pim/Structure/Integration/Attribute/Validation/BlacklistedAttributeCodeIntegration.php:34]: I'd also extract this bit for the sake of clarity

**Review by SamirBoulil — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Bundle/Resources/translations/validators.en_US.yml (hunk 3 above)
Comment (juliensnz) [reply src/Akeneo/Pim/Enrichment/Bundle/Resources/translations/validators.en_US.yml:73]: It's already planned but the proposed sentences are not valid so we will merge with this one

File: tests/back/Pim/Structure/Integration/Attribute/Query/IsAttributeCodeBlacklistedIntegration.php (hunk 4 above)
Comment (juliensnz) [reply tests/back/Pim/Structure/Integration/Attribute/Query/IsAttributeCodeBlacklistedIntegration.php:14]: ok!

**Pull request merged**